### PR TITLE
Version 1.2.0: Sponge Gourd

### PR DESCRIPTION
**Pull request by andtsa**
# Version 1.2.0: Sponge Gourd
tl;dr:
- `gourd analyse` redone, now supports:
    - CSV 
    - custom columns
    - custom metrics
    - grouping
- gourd.toml changes:
    - warn_on_label_overlap removed (pending merging of #41 )
    - label.priority is optional
- gourd set limits completely removed

## breaking changes:
- gourd analyse behaves very differently, check updated CLI documentation
- gourd set limits does not exist at all, functionality is integrated in gourd rerun. or just update gourd.toml and run again normally
- ...

## maintainer changes:
- can now create new metrics using `ColumnGenerator`(s)
- new integration test framework
- new pipeline with smaller actions for smaller PRs

## qol changes:
- pipeline artefacts expire every week

# still todo:
- [x] fix divide by zero error in default metrics accumulator
- [x] write new gourd analyse user-end documentation
- [x] write col gen maintainer docs
- [x] write config() integration test maintainer docs
- [x] create issues for additional integration test scenarios desired (multi-step pipelines, more extensive gourd analyse testing, etc)
- [x] update version in code & docs
- [x] merge new afterscript documentation
- [x] write new postprocessing documentation (using `next`) & tutorial
- [x] documentation for slurm.{mail_type, begin, mail_user, additional_args}
- [x] for every value in constants.rs, mention the lines in documentation that need to be updated when the constant changes.
- [x] merge #41 
- [x] improve help for `SLURM versioning failed` error in cli
- [x] remove mentions of set_limits in documentation
- [x] ...

- [x] after everything else is done, update `docs/maintainer/version-history/section.tex`

### CODE REVIEW

File: docs/maintainer/version-history/section.tex [hunk 1]
```diff
@@ -5,6 +5,14 @@ \section{Version History}
 \input{version-history/definitions}
 % \version{x} for start of version x section
 
+\version{1.2.0}{Sponge Gourd}{}
+
+Major internal reworkings, redesigned \texttt{gourd analyse}.
+Past this major release version the project is open source.
+For more details on this release, check out
```
Comment (mgazeel) [inline docs/maintainer/version-history/section.tex:12]: check out is not formal, use "see"

File: src/gourd/cli/def.rs [hunk 2]
```diff
@@ -165,47 +168,141 @@ pub struct AnalyseStruct {
...
-/// Arguments supplied with the `set-limits` command.
-#[derive(Args, Debug, Clone)]
-pub struct SetLimitsStruct {
-    /// The id of the experiment of which to change limits
-    /// [default: newest experiment]
-    #[arg(value_name = "EXPERIMENT")]
-    pub experiment_id: Option<usize>,
+/// Enum for subcommands of the `run` subcommand.
+#[derive(Subcommand, Debug, Clone)]
+pub enum AnalSubcommand {
```
Comment (mgazeel) [inline src/gourd/cli/def.rs:182]: Dont abbreviate Analyse

File: src/gourd/cli/process.rs [hunk 3]
```diff
@@ -196,29 +205,57 @@ pub async fn process_command(cmd: &Cli) -> Result<()> {
...
+                        {
+                            println!("{out}");
+                        } else {
+                            // INFO: here we can detect if the run is supposed to have an
+                            // afterscript ourselves. but for now the user can do it instead.
+                            bailc!(
+                                "could not print afterscript output",;
+                                "there is no afterscript output for this run",;
+                                "please make sure
+1. this run is supposed to have an afterscript, and
```
Comment (mgazeel) [inline src/gourd/cli/process.rs:236]: Broken indentation, you can use `\` to produce multi line strings

File: src/gourd/analyse/tests/mod.rs [hunk 4]
```diff
@@ -35,318 +24,67 @@ static TEST_RUSAGE: RUsage = RUsage {
...
-    let tmp_dir = TempDir::new("testing").unwrap();
-
-    let output_path = tmp_dir.path().join("");
-    let statuses = BTreeMap::new();
-
-    assert!(analysis_csv(&output_path, statuses).is_err());
-    assert!(tmp_dir.close().is_ok());
+    assert_eq!(
+        "
+| hello | world     |
```
Comment (mgazeel) [inline src/gourd/analyse/tests/mod.rs:39]: Fix indent!

File: src/gourd/analyse/tests/mod.rs [hunk 5]
```diff
@@ -35,318 +24,67 @@ static TEST_RUSAGE: RUsage = RUsage {
...
-        "second".to_string(),
-        vec![(0, 0), (1, 1), (2, 1), (3, 2), (4, 2), (5, 2)],
-    );
-
-    assert!(make_plot((5, 3, data), BitMapBackend::new(&output_path, (300, 300))).is_ok());
+        "
+| hello | hello     |
+*-------*-----------*
+| a     | a         |
+| hi    | b b b b b |
```
Comment (mgazeel) [inline src/gourd/analyse/tests/mod.rs:84]: Fix indent!

File: src/integration/versioning.rs [hunk 6]
```diff
@@ -1,42 +1,19 @@
...
-            afterscript: None,
-            resource_limits: None,
-            next: vec![],
-        },
-    );
-
-    save_gourd_toml(&conf, &env.temp_dir);
-    gourd!(env; "run", "local"; "failed to use repo versioning");
-}
+// use flate2::bufread::GzDecoder;
```
Comment (mgazeel) [inline src/integration/versioning.rs:1]: What is happening here?

File: src/integration/rerun.rs [hunk 7]
```diff
@@ -63,124 +44,111 @@ fn test_two_one_run() {
...
+//     let env = init();
+//     let (conf, conf_path) = config(&env,
+// "./src/integration/configurations/failing.toml").unwrap();
+//
+//     let experiment_path = conf.experiments_folder.join("1.lock");
+//     assert!(!experiment_path.exists());
+//
+//     let _ = gourd!(&env; "-c", conf_path.to_str().unwrap(), "run", "local";
+// "run local");
+//
```
[the remark this answers is not in the document]
Comment (andtsa) [reply src/integration/rerun.rs:62]: read comment above.

**Review by mgazeel — APPROVED** [review]

File: src/integration/versioning.rs (hunk 6 above)
Comment (andtsa) [reply src/integration/versioning.rs:1]: part of #31

File: src/integration/rerun.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (andtsa) [reply src/integration/rerun.rs:62]: tests fail due to a bug in faketty, a workaround isn't worth it so we will restore these tests when that lib gets updated
